adafruit_azureiot/device_registration.py

```diff
@@ -73,105 +62,102 @@ def __init__(
+            time.sleep(waittime)
+            self._operation_id = message["operationId"]
+        elif topic.startswith("$dps/registrations/res/200"):
+            self._hostname = message["registrationState"]["assignedHub"]
 
-        if data is not None and "status" in data:
-            if data["status"] == "assigning":
-                time.sleep(self._loop_interval)
-                if loop_try < 20:
-                    loop_try = loop_try + 1
-                    return self._loop_assign(operation_id, headers)
+    def _connect_to_mqtt(self) -> None:
+        self._mqtt.on_connect = self._on_connect
 
-                err = "ERROR: Unable to provision the device."
-                self._logger.error(err)
-                raise DeviceRegistrationError(err)
+        self._mqtt.connect()
 
-            if data["status"] == "assigned":
-                state = data["registrationState"]
-                return state["assignedHub"]
-        else:
-            data = str(data)
+        self._logger.info(
+            " - device_registration :: connect :: created mqtt client. connecting.."
+        )
+        while not self._auth_response_received:
+            self._mqtt.loop()
 
-        err = "DPS L => " + str(data)
-        self._logger.error(err)
-        raise DeviceRegistrationError(err)
+        self._logger.info(
+            f" - device_registration :: connect :: on_connect must be fired. Connected ? {str(self._mqtt_connected)}"
+        )
+
+        if not self._mqtt_connected:
+            raise DeviceRegistrationError("Cannot connect to MQTT")
+
+    def _start_registration(self) -> None:
+        self._mqtt.add_topic_callback(
+            "$dps/registrations/res/#", self._handle_dps_update
+        )
+        self._mqtt.subscribe("$dps/registrations/res/#")
+
+        message = json.dumps({"registrationId": self._device_id})
+
+        self._mqtt.publish(
+            f"$dps/registrations/PUT/iotdps-register/?$rid={self._device_id}", message
+        )
 
-    def _run_put_request_with_retry(self, url, body, headers):
         retry = 0
-        response = None
-
-        while True:
-            gc.collect()
-            try:
-                self._logger.debug("Trying to send...")
-                response = requests.put(url, json=body, headers=headers)
-                self._logger.debug("Sent!")
-                break
-            except RuntimeError as runtime_error:
-                self._logger.info(
-                    "Could not send data, retrying after 0.5 seconds: "
-                    + str(runtime_error)
-                )
-                retry = retry + 1
-
-                if retry >= 10:
-                    self._logger.error("Failed to send data")
-                    raise
-
-                time.sleep(0.5)
-                continue
-
-        gc.collect()
-        return response
-
-    def _run_get_request_with_retry(self, url, headers):
+
+        while self._operation_id is None and retry < 10:
```

Comment:
How does `retry` relate to `expiry` in `register_device`? Should retry loop's delay match?

Reply:
The retry is whilst it waits for a response to come back from the DPS - it can take a few seconds for the DPS to reply with the operation ID, and the final registration status.

The expiry is how long the tokens are valid for when connecting, so this is a lot longer (defaults to 6 hours IIRC)